touchforms/formplayer/experiments.py

```diff
@@ -0,0 +1,151 @@
+        action = request['action']
+        diff_logger.info(
+            "MSG",
+            extra={
+                'action': action,
+                'request': request,
+                'control': control_value,
+                'candidate': candidate_value
+            })
+
+    def log_timing(self, control, candidate):
+        timing_logger.info(
+            "MSG",
+            extra={
+                'action': self.name,
+                'control_duration': control.duration,
+                'candidate_duration': candidate.duration
+            })
+
+
+def compare_list(control, candidate):
+    is_equal = True
+    if len(control) != len(candidate):
+        logging.info("Length of control %s is different than candidate %s" % (control, candidate))
+    for first_item, second_item in zip(control, candidate):
+        if not formplayer_compare(first_item, second_item):
+            is_equal = False
+    return is_equal
+
+
+def compare_dict(control, candidate, current_key):
+    is_equal = True
+    for key in control:
+        if check_skip_key(key):
+            break
+        if key not in candidate:
+            logging.info('Key %s in control %s but not candidate %s' % (key, control, candidate))
+            is_equal = False
+        if not formplayer_compare(control.get(key), candidate.get(key), key):
+            is_equal = False
+    return is_equal
+
+
+def formplayer_compare(control, candidate, key=None):
+    if check_skip_key(key):
+        return True
+    if isinstance(control, dict):
+        if not(isinstance(candidate, dict)):
+            logging.info('For key %s: Candidate %s is not a dict, but control %s is' % (key, candidate, control))
+            return False
+        return compare_dict(control, candidate, key)
+    elif isinstance(control, list):
+        if not(isinstance(candidate, list)):
+            logging.info('For key %s: Candidate %s is not a list, but control %s is' % (key, candidate,
+                                                                                        control))
+            return False
+        return compare_list(control, candidate)
+    else:
+        return formplayer_string_compare(control, candidate, key)
+
+
+## Mappings between what Formplayer and Touchforms can safely disagree on
+def formplayer_string_compare(control, candidate, key=None):
+    if key == "repeatable":
+        # These end up as 0 and 1 in python world, despite being set to True and False in xformplayer.py
+        if control == 0:
+            ret = candidate == "false"
+        elif control == 1:
+            ret = candidate == "true"
+        else:
+            ret = control == candidate
+    else:
+        ret = control == candidate
+    if not ret:
+        logging.info("Mismatch with key %s between control %s and candidate %s"
+                     % (key, control, candidate))
+    return ret
+
+## Keys that SMS doesn't use or that we otherwise can skip
+def check_skip_key(key):
```

Comment:
@gcapalbo you might have done so on your first pass but would be great if you could double check the assumptions I made here (covered many of them on Slack)

Reply:
yes these are fine

Reply:
👍